### ISSUE TITLE
v1.1 Authentication in C#

### ISSUE THREAD

**Issue by jbasen** [OpenWonderLabs/SwitchBotAPI#137]
### Current Situation

I've had a working Crestron-SwitchBot driver written in C# with v1.0 of the SwitchBot protocol.  I'm working on converting that driver to v1.1 of the protocol because I want to include full support for the SwitchBot Lock including locking and unlocking the lock.  However, I'm running into a problem with authentication.  More specifically, I am receiving a return code of 401 when I try to perform a basic GET request for getting the device list..

I've tried to follow the JavaScript example code in my efforts to implement API v1.1 authentication but I'm not a JavaScript programmer.

Any ideas would be greatly appreciated.  Thanks in advance for the help

### Logs

```markdown
Relevant C# code running on the Crestron processor is as follows:

string token = "My Token";
string secret = "My Secret Key";
DateTime dt1970 = new DateTime(1970, 1, 1);
DateTime current = DateTime.Now;
TimeSpan span = current - dt1970;
long time = Convert.ToInt64(span.TotalMilliseconds);
string nonce = "requestID";
string data = token + time.ToString() + nonce;
Encoding utf8 = Encoding.UTF8;
HMACSHA256 hmac = new HMACSHA256(utf8.GetBytes(secret));
string signature = Convert.ToBase64String(hmac.ComputeHash(utf8.GetBytes(data)));

//Create url
string url = "https://api.switch-bot.com/v1.1/devices";

//Create http client
HttpsClient client = new HttpsClient();
HttpsClientRequest request = new HttpsClientRequest();
HttpsClientResponse response;

//Send http client request
request.KeepAlive = false;
request.Url.Parse(url);
request.RequestType = Crestron.SimplSharp.Net.Https.RequestType.Get;
request.Header.AddHeader(new HttpsHeader("Accept", "*/*"));
request.Header.AddHeader(new HttpsHeader("Authorization", token));
request.Header.AddHeader(new HttpsHeader("sign", signature));
request.Header.AddHeader(new HttpsHeader("nonce", nonce));
request.Header.AddHeader(new HttpsHeader("t", time.ToString()));

//Get response
response = client.Dispatch(request);
```

### Configuration

```json
N/A
```

### Environment

Crestron CP4 Processor
SwitchBot Hub Mini
SwitchBot Lock

**Comment by jbasen** [comment]
Still looking for a resolution.
Thanks

**Comment by asknoone** [comment]
Are you using that code verbatim? If so then it's probably the nonce of "requestID" causing the issue.

You need to generate a random string for the nonce, which is different for each request I believe.

I don't know why they didn't just include this as part of the example code as 99.9% of people using this integration don't care what the nonce is and won't have it generated by any other external system (which I think might be their rationale for not including it).

I haven't received my Switchbot yet so I haven't tried implementing any of this yet; but if / when I do get it working I'll add an example somewhere.

Unless someone else who has already done it can add it.

However, as soon as the Bluetooth API is working and added to Home Assistant I will be promptly going route instead and blocking my house door lock talking to the Internet :-)

**Comment by donavanbecker** [comment]
@asknoone, if you post an example here. I will add it to the documentation.

**Comment by jbasen** [comment]
Interesting. Thanks @asknoone !!  I’m on the road for a few days but will give that a try.  As @donavanbecker requested, an example would also help me.

thanks again

**Comment by hihouhou** [comment]
`require 'base64'
require 'openssl'
require 'uri'
require 'net/http'


mixed_token = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
p "token #{mixed_token}"
mixed_tokenbis = mixed_token.dup
tsbis = Time.now.to_i * 1000
mixed_tokenbis.concat("#{tsbis}")
p "mixedtokenbis #{mixed_tokenbis}"
sign = OpenSSL::HMAC.digest('sha256', 'YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY', mixed_tokenbis)
signbis = Base64.encode64(sign).gsub(/\n/, '')
p "sign #{signbis}"
p "ts #{ts}"
puts "curl -s 'https://api.switch-bot.com/v1.1/devices/ZZZZZZZZZZZZ/status' -H 'Authorization: #{mixed_token}' -H 'sign: #{signbis}' -H 'nonce: ' -H 't: #{tsbis}' -H 'Content-type: application/json' | jq ."
nonce = ''

uri = URI.parse("https://api.switch-bot.com/v1.1/devices/ZZZZZZZZZZZZ/status")
request = Net::HTTP::Get.new(uri)
request.content_type = "application/json"
request["Authorization"] = mixed_token
request["Sign"] = signbis
request["Nonce"] = nonce
request["T"] = tsbis

req_options = {
  use_ssl: uri.scheme == "https",
}

response = Net::HTTP.start(uri.hostname, uri.port, req_options) do |http|
  http.request(request)
end

p response.code
p response.body
`
weird problem for me (seems to be nonce the problem)
the generated version for curl works but the one with ruby is always 401....

**Comment by ytkg** [comment]
me too. the generated version for curl works but the one with ruby is always 401....

**Comment by jbasen** [comment]
I’m on the road right now or I would try this myself.  It would be interesting to point the code and curl generated calls to https://webhook.site and see what the differences are between the two calls.  I’ve found webhook.site to be a useful tool for this in the past.  I wish I had realized that a curl call would work when I was running into the same problem with my c# code.

**Comment by ytkg** [comment]
Oh! Thank you! I'll give it a try!

**Comment by hihouhou** [comment]
I didn't know https://webhook.site, very nice !!! thank you
https://webhook.site/#!/fcf4b5dc-986e-4c95-8c85-bf99c9233d4e/340f0077-02d4-45d1-a881-54ee0a884831/1 (my test)
with curl / ruby, I see nonce is removed for curl version

**Comment by hihouhou** [comment]
![Capture d’écran 2022-10-14 205422](https://user-images.githubusercontent.com/10907233/195920431-85e5aeb4-5327-4ae5-8ac5-9a9fdae5b05a.png)

**Comment by hihouhou** [comment]
```
require 'base64'
require 'openssl'
require 'uri'
require 'net/http'


mixed_token = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
p "token #{mixed_token}"
mixed_tokenbis = mixed_token.dup
tsbis = Time.now.to_i * 1000
mixed_tokenbis.concat("#{tsbis}")
p "mixedtokenbis #{mixed_tokenbis}"
sign = OpenSSL::HMAC.digest('sha256', 'YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY', mixed_tokenbis)
signbis = Base64.encode64(sign).gsub(/\n/, '')
p "sign #{signbis}"
#p "ts #{ts}"
nonce = ''
puts "curl -s 'https://webhook.site/fcf4b5dc-986e-4c95-8c85-bf99c9233d4e' -H 'Authorization: #{mixed_token}' -H 'sign: #{signbis}' -H 'nonce: ' -H 't: #{tsbis}' -H 'Content-type: application/json' | jq ."

uri = URI.parse("https://webhook.site/fcf4b5dc-986e-4c95-8c85-bf99c9233d4e")
request = Net::HTTP::Get.new(uri)
request.content_type = "application/json"
request["Authorization"] = mixed_token
request["sign"] = signbis
request["nonce"] = ''
request["t"] = tsbis.to_i


req_options = {
  use_ssl: uri.scheme == "https",
}

response = Net::HTTP.start(uri.hostname, uri.port, req_options) do |http|
  http.request(request)
end

p response.code
p response.body
```
it works with it !

**Comment by jbasen** [comment]
Since I’m going to have to translate this into c# andI have no experience with ruby, any explanation on what you’ve done to get it working would be greatly appreciated. Thanks

**Comment by hihouhou** [comment]
```
request["Sign"] = signbis -> request["sign"] = signbis
request["Nonce"] = nonce -> request["nonce"] = ''
request["T"] = tsbis -> request["t"] = tsbis.to_i
```
I hope it will help you

**Comment by hihouhou** [comment]
hum finally, random 200 and too many 401 :-(

**Comment by hihouhou** [comment]
I confirm after new test with https://www.uuidgenerator.net/dev-corner/ruby
25% -> 200 / 75% -> 401

**Comment by jbasen** [comment]
Thanks @hihouhou 

when I’m back in a week I’ll give your changes a try in c#.  Hopefully I can play around and get the changes to work 100% of the time. Thanks again

**Comment by hihouhou** [comment]
I'm sure the problem is on our side... I found example on internet with only the token for headers and my first try worked and after failed like when I tested with all mandatory headers ,it's disappointing and incomprehensible
![Capture d’écran 2022-10-15 112748](https://user-images.githubusercontent.com/10907233/195979328-5cb9a903-5f3b-4381-927d-8a32415db8a5.png)

**Comment by jbasen** [comment]
Thanks @hihouhou 

Do you have a timeframe for when this will be addressed?  I only ask because I have an article that includes the SwitchBot lock being published later in November.  My plan was to have my Crestron-SwitchBot driver updated to v1.1 of the SwitchBot API so readers could download the new driver from my GitHub.  Adding control of the SwitchBot lock to the driver requires v1.1 security.  In testing I found that without implementing v1.1 security I could only query the lock and door status; I can’t lock/unlock the lock. 

Thanks again for your help

**Comment by hihouhou** [comment]
the bash version for curl : 

bash /tmp/switchbot.sh <TOKEN> <SECRET>
```
#!/bin/bash

ts=$(date +%s%N)
token=$1
secret=$2
nonce=""
echo "${token}${ts}${nonce}"
sign=$(echo -en "${token}${ts}000${nonce}" | openssl dgst -sha256 -hmac "$secret" -binary | base64 | sed -e 's/+/-/g' -e 's/\//_/g')
echo "$sign"
curl -s "https://api.switch-bot.com/v1.1/devices" -H "Authorization: $token" -H "sign: $sign" -H "nonce: $nonce" -H "t: ${ts}000" -H "Content-type: application/json"

```

**Comment by asknoone** [comment]
My Switchbot hasn't arrived yet so I can't try it myself, but based on the comments above it still looks to me like you are not changing the nonce for each request.

You need to generate a unique GUID for each request and use that as the nonce. It's probably 1 or two lines of code to generate a GUID.

I imagine your first request worked because you had the nonce as in the example "RequestID"? And therefore that was a unique string the first time you did a request. But subsequent ones, it would go "oh no, seen this before so I'm going to bin the request". It helps prevent replay attacks.

If you occasionally get 200s then it's probably one of these 2 reasons (these are guesses, I have no idea how their API is coded):
1. They may have a duration after which a nonce can be reused
2. They have a load balanced API service and they only validate to see if the nonce has been seen before per worker and therefore if you hit another one if their workers with the same nonce then it would be allowed.

**Comment by jbasen** [comment]
Thanks.  I have all the coding changes done and will give it a try when I'm back from my trip in a few days

**Comment by hihouhou** [comment]
> My Switchbot hasn't arrived yet so I can't try it myself, but based on the comments above it still looks to me like you are not changing the nonce for each request.
> 
> You need to generate a unique GUID for each request and use that as the nonce. It's probably 1 or two lines of code to generate a GUID.
> 
> I imagine your first request worked because you had the nonce as in the example "RequestID"? And therefore that was a unique string the first time you did a request. But subsequent ones, it would go "oh no, seen this before so I'm going to bin the request". It helps prevent replay attacks.
> 
> If you occasionally get 200s then it's probably one of these 2 reasons (these are guesses, I have no idea how their API is coded):
> 
>     1. They may have a duration after which a nonce can be reused
> 
>     2. They have a load balanced API service and they only validate to see if the nonce has been seen before per worker and therefore if you hit another one if their workers with the same nonce then it would be allowed.

look at my script with bash  and no change for nonce it works, like on : 
https://github.com/OpenWonderLabs/SwitchBotAPI/blob/main/README.md#sending-a-request , for Python 2 example code and Python 3 example code, it seems to be empty too, my problem is only for ruby (my first goal :-) )

**Comment by asknoone** [comment]
I can't figure it out.

The examples are inconsistent.

The "for instance" example shows just signing with token + 13 digit timestamp.
The Python 2 and 3 example shows token + timestamp + nonce. (The nonce is blank in the example but presumably you need to use a uuid)
The JavaScript example shows token + timestamp + nonce (The nonce is populated with "requestID" in this example, again, presumably you need to use a uuid)

I've been trying to get it to work in NodeRed.

**Comment by hihouhou** [comment]
interesting thing! when I keep the same user agent , no more problem with curl and ruby code

**Comment by ytkg** [comment]
@hihouhou 
What user agent did you set?
Also, I would like to see some sample code.

**Comment by hihouhou** [comment]
> @hihouhou What user agent did you set? Also, I would like to see some sample code.

per default https://github.com/OpenWonderLabs/SwitchBotAPI/issues/137#issuecomment-1279352902 but now : 
for curl -H 'User-Agent: Switchbot (https://github.com/hihouhou/huginn_switchbot_agent)'
for ruby request["User-Agent"] = " Switchbot (https://github.com/hihouhou/huginn_switchbot_agent)"

**Comment by ytkg** [comment]
Thank you.
I'll give it a try.

**Comment by ytkg** [comment]
Hmmm, it doesn't work in my environment.

**Comment by jbasen** [comment]
I am back from my trip so I've been trying everything suggested in this thread without success.  

My code (see below) now uses a Guid for the nonce.  I have tried various header changes and have tried not setting a nonce header.  

Though I'm not a bash programmer I did research on the code written by @hihouhou. I did notice one thing.  I believe the statement

ts=$(date +%s%N)

sets ts with the time in nanoseconds, not milliseconds.  So this will be longer than the 13 digit millisecond time that SwitchBot is expecting.  So, when the token is concatenated with the time and the nonce is set to "" in the bash script, if SwitchBot expects a fixed length token and 13 digits for the time, then the bash script is actually concatenating on extra digits that the system would interpret as nonce.  

Below is the current state of my code that still gets me a 401 response every time I've run it.  As others have said, the inconsistency of the examples has made this very challenging.  Any additional ideas / solutions would be greatly appreciated.

_____________________________

string token = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX";
string secret = "YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY";
DateTime dt1970 = new DateTime(1970, 1, 1);
dt1970 = DateTime.SpecifyKind(dt1970, DateTimeKind.Utc);
DateTime current = DateTime.Now;
current = DateTime.SpecifyKind(current, DateTimeKind.Utc);
TimeSpan span = current - dt1970;
long time = Convert.ToInt64(span.TotalMilliseconds);
Guid g = Guid.NewGuid();
string nonce = g.ToString();
string data = token + time.ToString() + nonce;
Encoding utf8 = Encoding.UTF8;
HMACSHA256 hmac = new HMACSHA256(utf8.GetBytes(secret));
string signature = Convert.ToBase64String(hmac.ComputeHash(utf8.GetBytes(data)));

//Create url
string url = "https://api.switch-bot.com/v1.1/devices";

//Create http client
HttpsClient client = new HttpsClient();
client.HostVerification = false;
client.PeerVerification = false;

//client.Verbose = true;
HttpsClientRequest request = new HttpsClientRequest();

HttpsClientResponse response;

//Send http client request
request.KeepAlive = false;
request.Url.Parse(url);
request.RequestType = Crestron.SimplSharp.Net.Https.RequestType.Get;
request.Header.AddHeader(new HttpsHeader("accept", "*/*"));
request.Header.AddHeader(new HttpsHeader("authorization", token));
request.Header.AddHeader(new HttpsHeader("sign", signature));
request.Header.AddHeader(new HttpsHeader("nonce", nonce));
request.Header.AddHeader(new HttpsHeader("t", time.ToString()));
request.Header.AddHeader(new HttpsHeader("content-type", "application/json"));
request.Header.AddHeader(new HttpsHeader("'User-Agent", "Crestron"));

**Comment by hihouhou** [comment]
actually I cheat like that -> https://github.com/hihouhou/huginn_switchbot_agent/blob/main/lib/huginn_switchbot_agent/switchbot_agent.rb#L163 (not proud about what I did)

**Comment by jbasen** [comment]
Well.  I took the bash script, cleaned up the "cheats", plugged in my values for token and secret, captured the values for time and nonce from a run of my code, and sent the output to webhook.site.  With this I have been able to validate that my code is producing the exact headers and the exact same encrypted value for sign.  I've also removed the setting of a user agent header in my code as that didn't seem to work properly when I sent the https request to webhook.site to test it.

So, I'm really stuck at this point as my code seems to be doing everything correctly but I still get a 401 response :-(

**Comment by Write** [comment]
Trying curl with no success too.; It just says "{"message":"Authorization header requires 'Credential' parameter. Authorization header requires 'Signature' parameter. Authorization header requires 'SignedHeaders' parameter. Authorization header requires existence of either a 'X-Amz-Date' or a 'Date' header. Authorization=[redacted]"}%"

**Comment by ytkg** [comment]
I noticed that HTTP 1.1 also fails with curl.

I was able to confirm that using HTTP 2 works with Ruby.
```ruby
require 'securerandom'
require 'base64'
require 'openssl'
require 'net-http2'

token = 'your token'
secret = 'your secret'
t = (Time.now.to_f * 1000).to_i
nonce = SecureRandom.alphanumeric

sign = Base64.strict_encode64(OpenSSL::HMAC.digest('sha256', secret, "#{token}#{t}#{nonce}"))

puts "Authorization: #{token}"
puts "t: #{t}"
puts "sign: #{sign}"
puts "nonce: #{nonce}"

headers = {
  'Authorization' => token,
  't' => t,
  'sign' => sign,
  'nonce' => nonce,
}

client = NetHttp2::Client.new("https://api.switch-bot.com")
response = client.call(:get, '/v1.1/devices', headers: headers)

puts response.body
```

**Comment by hihouhou** [comment]
> I noticed that curl also fails when using HTTP 1.1.
> 
> I was able to confirm that using HTTP 2 works with Ruby.
> 
> ```ruby
> require 'securerandom'
> require 'base64'
> require 'openssl'
> require 'net-http2'
> 
> token = 'your token'
> secret = 'your secret'
> t = (Time.now.to_f * 1000).to_i
> nonce = SecureRandom.alphanumeric
> 
> sign = Base64.strict_encode64(OpenSSL::HMAC.digest('sha256', secret, "#{token}#{t}#{nonce}"))
> 
> puts "Authorization: #{token}"
> puts "t: #{t}"
> puts "sign: #{sign}"
> puts "nonce: #{nonce}"
> 
> headers = {
>   'Authorization' => token,
>   't' => t,
>   'sign' => sign,
>   'nonce' => nonce,
> }
> 
> client = NetHttp2::Client.new("https://api.switch-bot.com")
> response = client.call(:get, '/v1.1/devices', headers: headers)
> 
> puts response.body
> ```

https://github.com/OpenWonderLabs/SwitchBotAPI/issues/137#issuecomment-1288141785

**Comment by Write** [comment]
> I noticed that curl also fails when using HTTP 1.1.
> 
> I was able to confirm that using HTTP 2 works with Ruby.
> 
> ```ruby
> require 'securerandom'
> require 'base64'
> require 'openssl'
> require 'net-http2'
> 
> token = 'your token'
> secret = 'your secret'
> t = (Time.now.to_f * 1000).to_i
> nonce = SecureRandom.alphanumeric
> 
> sign = Base64.strict_encode64(OpenSSL::HMAC.digest('sha256', secret, "#{token}#{t}#{nonce}"))
> 
> puts "Authorization: #{token}"
> puts "t: #{t}"
> puts "sign: #{sign}"
> puts "nonce: #{nonce}"
> 
> headers = {
>   'Authorization' => token,
>   't' => t,
>   'sign' => sign,
>   'nonce' => nonce,
> }
> 
> client = NetHttp2::Client.new("https://api.switch-bot.com")
> response = client.call(:get, '/v1.1/devices', headers: headers)
> 
> puts response.body
> ```

Doesn't change a thing for me, using --http2 with curl it throws randomly Unauthorized for no reasons at all.

**Comment by iCarrot0605** [comment]
I used Guid for nonce.This code works my environment.
Because I am beginner for C#, there may be glitches in your environment. 

```c#
using System;
using System.Diagnostics;
using System.Text;
using System.Security.Cryptography;
using System.Net.Http;

string token = "My Token";
string secret = "My Secret Key";
DateTime dt1970 = new DateTime(1970, 1, 1);
DateTime current = DateTime.Now;
TimeSpan span = current - dt1970;
long time = Convert.ToInt64(span.TotalMilliseconds);
string nonce = Guid.NewGuid().ToString();
string data = token + time.ToString() + nonce;
Encoding utf8 = Encoding.UTF8;
HMACSHA256 hmac = new HMACSHA256(utf8.GetBytes(secret));
string signature = Convert.ToBase64String(hmac.ComputeHash(utf8.GetBytes(data)));

//Create http client
HttpClient client = new HttpClient();
var request = new HttpRequestMessage(HttpMethod.Get, @"https://api.switch-bot.com/v1.1/devices");
request.Headers.TryAddWithoutValidation(@"Authorization", token);
request.Headers.TryAddWithoutValidation(@"sign", signature);
request.Headers.TryAddWithoutValidation(@"nonce", nonce);
request.Headers.TryAddWithoutValidation(@"t", time.ToString());

var response = await client.SendAsync(request);

Console.WriteLine(await response.Content.ReadAsStringAsync());
```

**Comment by jbasen** [comment]
Thanks @iCarrot0605 !!

**Comment by alamers** [comment]
For the Java fans out there, I translated the c# example to Java (11+):

```
    public static void main(String[] args) throws Exception {
    	
      String token = args[0];
      String secret = args[1];
      String nonce = UUID.randomUUID().toString();
      String time= "" + Instant.now().toEpochMilli();
      String data = token + time + nonce;
      
      SecretKeySpec secretKeySpec = new SecretKeySpec(secret.getBytes("UTF-8"), "HmacSHA256");
      Mac mac = Mac.getInstance("HmacSHA256");
      mac.init(secretKeySpec);
      String signature = new String(Base64.getEncoder().encode(mac.doFinal(data.getBytes("UTF-8"))));      
      
      HttpRequest getDevices = HttpRequest.newBuilder()
      .uri(new URI("https://api.switch-bot.com/v1.1/devices"))
      .header("Authorization", token)
      .header("sign", signature)
      .header("nonce", nonce)
      .header("t", time)
      .GET()
      .build();
      
      HttpResponse<String> response = HttpClient.newBuilder().build().send(getDevices, BodyHandlers.ofString());
      
      System.out.println(response.body());
    	
    	
    }
```

**Comment by donavanbecker** [comment]
> For the Java fans out there, I translated the c# example to Java (11+):
> 
> 
> 
> ```
> 
>     public static void main(String[] args) throws Exception {
> 
>     	
> 
>       String token = args[0];
> 
>       String secret = args[1];
> 
>       String nonce = UUID.randomUUID().toString();
> 
>       String time= "" + Instant.now().toEpochMilli();
> 
>       String data = token + time + nonce;
> 
>       
> 
>       SecretKeySpec secretKeySpec = new SecretKeySpec(secret.getBytes("UTF-8"), "HmacSHA256");
> 
>       Mac mac = Mac.getInstance("HmacSHA256");
> 
>       mac.init(secretKeySpec);
> 
>       String signature = new String(Base64.getEncoder().encode(mac.doFinal(data.getBytes("UTF-8"))));      
> 
>       
> 
>       HttpRequest getDevices = HttpRequest.newBuilder()
> 
>       .uri(new URI("https://api.switch-bot.com/v1.1/devices"))
> 
>       .header("Authorization", token)
> 
>       .header("sign", signature)
> 
>       .header("nonce", nonce)
> 
>       .header("t", time)
> 
>       .GET()
> 
>       .build();
> 
>       
> 
>       HttpResponse<String> response = HttpClient.newBuilder().build().send(getDevices, BodyHandlers.ofString());
> 
>       
> 
>       System.out.println(response.body());
> 
>     	
> 
>     	
> 
>     }
> 
> ```

Can you add this to the readme with a PR?

**Comment by hihouhou** [comment]
> I noticed that curl also fails when using HTTP 1.1.
> 
> I was able to confirm that using HTTP 2 works with Ruby.
> 
> ```ruby
> require 'securerandom'
> require 'base64'
> require 'openssl'
> require 'net-http2'
> 
> token = 'your token'
> secret = 'your secret'
> t = (Time.now.to_f * 1000).to_i
> nonce = SecureRandom.alphanumeric
> 
> sign = Base64.strict_encode64(OpenSSL::HMAC.digest('sha256', secret, "#{token}#{t}#{nonce}"))
> 
> puts "Authorization: #{token}"
> puts "t: #{t}"
> puts "sign: #{sign}"
> puts "nonce: #{nonce}"
> 
> headers = {
>   'Authorization' => token,
>   't' => t,
>   'sign' => sign,
>   'nonce' => nonce,
> }
> 
> client = NetHttp2::Client.new("https://api.switch-bot.com")
> response = client.call(:get, '/v1.1/devices', headers: headers)
> 
> puts response.body
> ```

https://github.com/OpenWonderLabs/SwitchBotAPI/issues/137#issuecomment-1288141785

> ```ruby
> net-http2
> ```

I confirm that it works! thank you but why? is it possible to have an official explanation?

**Comment by donavanbecker** [comment]
@ytkg nope I do not.

**Comment by jbasen** [comment]
@iCarrot0605 - Would it be possible for you to post the version of .NET Framework, .Net Core (5,6,7?), etc.  that you used when testing the C# code you posted above that you said worked.  l am trying to track down what could be different between your environment and mine to try and figure out why my code still won't work.  

Thanks in advance for the help

**Comment by iCarrot0605** [comment]
@jbasen My environment is .NET 6.0.

**Comment by jbasen** [comment]
Thanks @iCarrot0605

**Comment by jbasen** [comment]
Hi @iCarrot0605 

I am working with Crestron tech support trying to get this to work.  We are trying to simply run the code you posted and we are unable to get your code to pass authentication.  

I'm not sure if you have ever used https://webhook.site/ You can point an http or https call there and it will display all the headers, etc. I use it all the time when I'm writing drivers to see exactly what is happening when, for example, the only example I have of how to call a driver is in curl. 

Could you please point your C# code to https://webhook.site/{the unique ID you get when you go to the site} instead of https://api.switch-bot.com/v1.1/... and then post all the information about your https call back here.  That may give us a clue as to what is working differently when you run your own code vs. when we run it. 

Thanks so much for your help

**Comment by iCarrot0605** [comment]
@jbasen 
In my environment, result is here.

```plain


Headers
--
connection | close
t | 1668243344605
nonce | abf6df7c-3b51-40ac-bcad-f1718591c501
sign | eynQFMB/lNdnopw0I/3YGxnJxRbhIKRGn8flqf+ukwQ=
authorization | XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
host | webhook.site
content-length |  
content-type |  

```

**Comment by jbasen** [comment]
Thanks @iCarrot0605

**Comment by jbasen** [comment]
Hi @iCarrot0605 

There is nothing in the headers from your request that looks different than what my code is sending to the API.  

However, I have a thought.  Are you a SwitchBot employee or in some way associated with SwitchBot?  If so, is there a possibility that your token and secret key being associated with an employee's account have a different security privilege than I do?  I'm just "grasping at straws" to try and understand why when I use your code with my token and secret key I get a 401 error and you are saying that the code works for you when you use your security credentials.  

Thanks again for your help

**Comment by iCarrot0605** [comment]
Hi @jbasen 
I'm not Switchbot employee.
I hope your program works well.

**Comment by jbasen** [comment]
Thanks @iCarrot0605

**Comment by jbasen** [comment]
The problem I was having with the v1.1 security API is now resolved.  I have to hand it to Crestron Tech Support that figured out the solution.  It all had to do with the Timestamp calculation I was using (which I actually took off of StackOverflow since everything posted on the Internet must be true).  

My original calculation of:

DateTime dt1970 = new DateTime(1970, 1, 1);
DateTime current = DateTime.Now;
TimeSpan span = current - dt1970;
long time = Convert.ToInt64(span.TotalMilliseconds);

produces a nice 13 digit value that looks valid but isn't.  The correct calculation is:

DateTime current = DateTime.UtcNow;
var dt1970 = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
var time = current.ToUniversalTime().Subtract(dt1970).TotalMilliseconds;

Thanks to everyone for their help
